File: packages/synthsr/src/volume.js
// SynthSR's spatial/intensity contract, independent of UI and inference runtime.
// Ported from neurolabusc/py_synthsr (Apache-2.0); see NOTICE.
import * as nifti from 'nifti-reader-js';

const product = (dims) => dims.reduce((a, b) => a * b, 1);
const index = (x, y, z, d) => x + d[0] * (y + d[1] * z);

export function readVolume(buffer) {
  if (nifti.isCompressed(buffer)) buffer = nifti.decompress(buffer);
  if (!nifti.isNIFTI(buffer)) throw new Error('Choose a NIfTI image (.nii or .nii.gz).');
  const h = nifti.readHeader(buffer);
  if (h.dims.slice(4, h.dims[0] + 1).some((d) => d > 1)) throw new Error('SynthSR needs a single 3D image. Extract a volume from this 4D image first.');
  const dims = h.dims.slice(1, 4);
  if (dims.some((d) => !Number.isInteger(d) || d < 2) || product(dims) > 128 * 1024 * 1024) throw new Error('Unsupported image dimensions.');
  const raw = nifti.readImage(h, buffer);
  const types = { 2: ['getUint8', 1], 4: ['getInt16', 2], 8: ['getInt32', 4], 16: ['getFloat32', 4], 64: ['getFloat64', 8], 256: ['getInt8', 1], 512: ['getUint16', 2], 768: ['getUint32', 4] };
  const type = types[h.datatypeCode];
  if (!type) throw new Error(`Unsupported NIfTI datatype ${h.datatypeCode}. Use a scalar intensity image.`);
  if (raw.byteLength < product(dims) * type[1]) throw new Error('The NIfTI voxel data is truncated.');
  const data = new Float32Array(product(dims));
  const view = new DataView(raw);
  const slope = h.scl_slope || 1, intercept = h.scl_slope ? h.scl_inter : 0;
  for (let i = 0; i < data.length; i++) {
    data[i] = view[type[0]](i * type[1], h.littleEndian) * slope + intercept;
    if (!Number.isFinite(data[i])) throw new Error('The image contains non-finite intensities. Clean NaN/Infinity values before synthesis.');
  }
  const unit = h.xyzt_units & 7;
  const scale = unit === 1 ? 1000 : unit === 3 ? 0.001 : 1;
  const affine = h.affine.map((row, r) => row.map((v) => r < 3 ? v * scale : v));
  if (affine.flat().some((v) => !Number.isFinite(v))) throw new Error('Invalid NIfTI affine.');
  inverse3(affine);
  return { data, dims, affine };
}

export function writeVolume({ data, dims, affine }, description = 'SynthSR synthetic T1; 1mm; not acquired T1') {
  const float = data instanceof Float32Array;
  const buffer = new ArrayBuffer(352 + data.length * (float ? 4 : 1));
  const v = new DataView(buffer);
  v.setInt32(0, 348, true); v.setInt16(40, 3, true);
  for (let a = 0; a < 7; a++) v.setInt16(42 + a * 2, dims[a] || 1, true);
  v.setInt16(70, float ? 16 : 2, true); v.setInt16(72, float ? 32 : 8, true);
  v.setFloat32(76, 1, true);
  for (let a = 0; a < 3; a++) v.setFloat32(80 + a * 4, Math.hypot(...affine.slice(0, 3).map((r) => r[a])), true);
  v.setFloat32(108, 352, true); v.setFloat32(112, 1, true); v.setUint8(123, 2);
  v.setInt16(254, 1, true);
  for (let r = 0; r < 3; r++) for (let c = 0; c < 4; c++) v.setFloat32(280 + 16 * r + 4 * c, affine[r][c], true);
  new Uint8Array(buffer, 148, 80).set(new TextEncoder().encode(description).slice(0, 79));
  new Uint8Array(buffer, 344, 4).set([110, 43, 49, 0]);
  if (float) for (let i = 0; i < data.length; i++) v.setFloat32(352 + i * 4, data[i], true);
  else new Uint8Array(buffer, 352).set(data);
  return buffer;
}

function inverse3(m) {
  const [[a,b,c], [d,e,f], [g,h,i]] = m;
  const det = a*(e*i-f*h)-b*(d*i-f*g)+c*(d*h-e*g);
  if (Math.abs(det) < 1e-12) throw new Error('The image affine is singular.');
  return [[e*i-f*h,c*h-b*i,b*f-c*e],[f*g-d*i,a*i-c*g,c*d-a*f],[d*h-e*g,b*g-a*h,a*e-b*d]].map((r) => r.map((v) => v/det));
}

export function rasAxes(affine) {
  const inverse = inverse3(affine);
  const axes = [0,1,2].map((c) => [0,1,2].reduce((a,b) => Math.abs(inverse[a][c]) >= Math.abs(inverse[b][c]) ? a : b));
  for (let i = 0; i < 3; i++) if (!axes.includes(i)) {
    const counts = [0,1,2].map((v) => axes.filter((a) => a === v).length);
    axes[axes.lastIndexOf(counts.indexOf(Math.max(...counts)))] = i;
  }
  return axes;
}

function reflect(i, n) {
  const p = ((i % (2*n)) + 2*n) % (2*n);
  return p < n ? p : 2*n-1-p;
}

// scipy.ndimage.gaussian_filter: separable, reflect boundaries, truncate=4.
export function gaussian(data, dims, sigmas) {
  let current = data;
  for (let axis = 0; axis < 3; axis++) {
    const sigma = sigmas[axis], radius = Math.floor(4*sigma + 0.5);
    if (sigma < 1e-15 || radius === 0) continue;
    const kernel = Array.from({ length: 2*radius+1 }, (_,i) => Math.exp(-0.5*((i-radius)/sigma)**2));
    const sum = kernel.reduce((a,b) => a+b, 0);
    const out = new Float32Array(data.length);
    const stride = axis === 0 ? 1 : axis === 1 ? dims[0] : dims[0]*dims[1];
    for (let z=0; z<dims[2]; z++) for (let y=0; y<dims[1]; y++) for (let x=0; x<dims[0]; x++) {
      const pos = axis === 0 ? x : axis === 1 ? y : z, base = index(x,y,z,dims) - pos*stride;
      let value = 0;
      for (let k=-radius; k<=radius; k++) value += current[base+reflect(pos+k,dims[axis])*stride]*kernel[k+radius]/sum;
      out[index(x,y,z,dims)] = value;
    }
    current = out;
  }
  return current;
}

export function resample1mm(volume) {
  const { data, dims, affine } = volume;
  const factors = [0,1,2].map((a) => Math.sqrt(affine.slice(0,3).reduce((sum,r) => sum+r[a]*r[a],0)));
  if (factors.some((s) => s < 0.05 || s > 20)) throw new Error('Voxel spacing is outside the supported range (0.05–20 mm).');
  // Match NumPy arange(start, stop, step), including its endpoint rounding.
  const starts=factors.map((f)=>-(f-1)/(2*f)), steps=factors.map((f)=>1/f);
  const shape=dims.map((d,a)=>Math.ceil(((starts[a]+steps[a]*Math.ceil(d*factors[a]))-starts[a])/steps[a]));
  if (product(shape) > 48*1024*1024) throw new Error('The 1 mm output is too large. Crop excess background before synthesis.');
  const filtered = gaussian(data, dims, factors.map((f) => f > 1 ? 0 : 0.25/f));
  const coords = shape.map((n,a) => Array.from({length:n}, (_,i) => Math.min(dims[a]-1, Math.max(0,starts[a]+i*((starts[a]+steps[a])-starts[a])))));
  const out = new Float32Array(product(shape));
  for (let z=0; z<shape[2]; z++) {
    const fz=coords[2][z], z0=Math.floor(fz), z1=Math.min(z0+1,dims[2]-1), tz=fz-z0;
    for (let y=0; y<shape[1]; y++) {
      const fy=coords[1][y], y0=Math.floor(fy), y1=Math.min(y0+1,dims[1]-1), ty=fy-y0;
      const b00=dims[0]*(y0+dims[1]*z0), b10=dims[0]*(y1+dims[1]*z0);
      const b01=dims[0]*(y0+dims[1]*z1), b11=dims[0]*(y1+dims[1]*z1);
      for (let x=0; x<shape[0]; x++) {
        const fx=coords[0][x], x0=Math.floor(fx), x1=Math.min(x0+1,dims[0]-1), tx=fx-x0;
        const v00=filtered[b00+x0]*(1-tx)+filtered[b00+x1]*tx;
        const v10=filtered[b10+x0]*(1-tx)+filtered[b10+x1]*tx;
        const v01=filtered[b01+x0]*(1-tx)+filtered[b01+x1]*tx;
        const v11=filtered[b11+x0]*(1-tx)+filtered[b11+x1]*tx;
        out[index(x,y,z,shape)]=(v00*(1-ty)+v10*ty)*(1-tz)+(v01*(1-ty)+v11*ty)*tz;
      }
    }
  }
  const aff = affine.map((r) => [...r]);
  for (let r=0;r<3;r++) {
    for (let a=0;a<3;a++) aff[r][a] /= factors[a];
    aff[r][3] -= [0,1,2].reduce((v,a) => v+aff[r][a]*0.5*(factors[a]-1),0);
  }
  return {data:out,dims:shape,affine:aff};
}

export function prepare(volume, { ct = false } = {}) {
  const resampled = resample1mm(volume);
  const axes = rasAxes(resampled.affine);
  const flips = axes.map((a,r) => resampled.affine[r][a] < 0);
  const alignedDims = axes.map((a) => resampled.dims[a]);
  const paddedDims = alignedDims.map((s) => Math.ceil(s/32)*32);
  const offsets = paddedDims.map((s,a) => Math.floor((s-alignedDims[a])/2));
  const input = new Float32Array(product(paddedDims));
  let min = Infinity, max = -Infinity;
  for (let x=0;x<alignedDims[0];x++) for (let y=0;y<alignedDims[1];y++) for (let z=0;z<alignedDims[2];z++) {
    const src = [0,0,0];
    [x,y,z].forEach((p,a) => { src[axes[a]]=flips[a]?alignedDims[a]-1-p:p; });
    let v=resampled.data[index(...src,resampled.dims)];
    if(ct) v=Math.min(80,Math.max(0,v));
    const j=((x+offsets[0])*paddedDims[1]+y+offsets[1])*paddedDims[2]+z+offsets[2];
    input[j]=v; min=Math.min(min,v); max=Math.max(max,v);
  }
  if (max <= min) throw new Error('The input has no intensity variation after preprocessing.');
  if (input.length > resampled.data.length) { min=Math.min(min,0); max=Math.max(max,0); }
  for(let i=0;i<input.length;i++) input[i]=(input[i]-min)/(max-min);
  return { input, paddedDims, alignedDims, offsets, axes, flips, dims:resampled.dims, affine:resampled.affine };
}

export function flipInput(input, dims) {
  const out=new Float32Array(input.length), slab=dims[1]*dims[2];
  for(let x=0;x<dims[0];x++) out.set(input.subarray(x*slab,(x+1)*slab),(dims[0]-1-x)*slab);
  return out;
}

export function finish(prediction, prep, { sharpen = true } = {}) {
  const {alignedDims:d,paddedDims:p,offsets:o,axes,flips,dims,affine}=prep;
  const aligned=new Float32Array(product(d));
  for(let x=0;x<d[0];x++) for(let y=0;y<d[1];y++) for(let z=0;z<d[2];z++) aligned[index(x,y,z,d)]=prediction[((x+o[0])*p[1]+y+o[1])*p[2]+z+o[2]];
  if(sharpen) {
    const smooth=gaussian(aligned,d,[1.5,1.5,1.5]);
    for(let i=0;i<aligned.length;i++) aligned[i]=2*aligned[i]-smooth[i];
  }
  const data=new Uint8Array(product(dims));
  for(let x=0;x<d[0];x++) for(let y=0;y<d[1];y++) for(let z=0;z<d[2];z++) {
    const dst=[0,0,0]; [x,y,z].forEach((v,a)=>{dst[axes[a]]=flips[a]?d[a]-1-v:v;});
    // numpy uint8 conversion truncates, it does not round.
    data[index(...dst,dims)]=Math.max(0,Math.min(255,2*aligned[index(x,y,z,d)]));
  }
  return {data,dims,affine};
}

File: exes/synthsr/src/volume.rs
// ponytail: loops mirror the JS port index for index; clamp differs from max/min only for NaN, which is rejected earlier.
#![allow(clippy::needless_range_loop, clippy::manual_clamp)]
// SynthSR spatial/intensity contract, ported from packages/synthsr/src/volume.js.
// Arithmetic is f64 with f32 storage at the same points as the JS, so results match bit for bit.
use crate::nifti::{product, Affine, Volume};

#[inline]
fn index(x: usize, y: usize, z: usize, d: &[usize; 3]) -> usize {
    x + d[0] * (y + d[1] * z)
}

pub fn inverse3(m: &Affine) -> Result<[[f64; 3]; 3], String> {
    let [[a, b, c, _], [d, e, f, _], [g, h, i, _]] = *m;
    let det = a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g);
    if det.abs() < 1e-12 {
        return Err("The image affine is singular.".into());
    }
    let adj = [
        [e * i - f * h, c * h - b * i, b * f - c * e],
        [f * g - d * i, a * i - c * g, c * d - a * f],
        [d * h - e * g, b * g - a * h, a * e - b * d],
    ];
    Ok(adj.map(|r| r.map(|v| v / det)))
}

pub fn ras_axes(affine: &Affine) -> Result<[usize; 3], String> {
    let inv = inverse3(affine)?;
    let mut axes = [0usize; 3];
    for c in 0..3 {
        axes[c] = (0..3)
            .reduce(|a, b| {
                if inv[a][c].abs() >= inv[b][c].abs() {
                    a
                } else {
                    b
                }
            })
            .unwrap();
    }
    for i in 0..3 {
        if !axes.contains(&i) {
            let counts: Vec<usize> = (0..3)
                .map(|v| axes.iter().filter(|&&a| a == v).count())
                .collect();
            let most = counts
                .iter()
                .position(|&c| c == *counts.iter().max().unwrap())
                .unwrap();
            let last = axes.iter().rposition(|&a| a == most).unwrap();
            axes[last] = i;
        }
    }
    Ok(axes)
}

fn reflect(i: isize, n: isize) -> usize {
    let p = ((i % (2 * n)) + 2 * n) % (2 * n);
    (if p < n { p } else { 2 * n - 1 - p }) as usize
}

// scipy.ndimage.gaussian_filter: separable, reflect boundaries, truncate=4.
pub fn gaussian(data: &[f32], dims: &[usize; 3], sigmas: [f64; 3]) -> Vec<f32> {
    let mut current: Vec<f32> = data.to_vec();
    for axis in 0..3 {
        let sigma = sigmas[axis];
        let radius = (4.0 * sigma + 0.5).floor() as isize;
        if sigma < 1e-15 || radius == 0 {
            continue;
        }
        let kernel: Vec<f64> = (0..=2 * radius)
            .map(|i| (-0.5 * ((i - radius) as f64 / sigma).powi(2)).exp())
            .collect();
        let sum: f64 = kernel.iter().sum();
        let mut out = vec![0f32; data.len()];
        let stride = [1, dims[0], dims[0] * dims[1]][axis];
        let n = dims[axis] as isize;
        for z in 0..dims[2] {
            for y in 0..dims[1] {
                for x in 0..dims[0] {
                    let pos = [x, y, z][axis];
                    let base = index(x, y, z, dims) - pos * stride;
                    let mut value = 0.0f64;
                    for k in -radius..=radius {
                        value += current[base + reflect(pos as isize + k, n) * stride] as f64
                            * kernel[(k + radius) as usize]
                            / sum;
                    }
                    out[index(x, y, z, dims)] = value as f32;
                }
            }
        }
        current = out;
    }
    current
}

pub fn resample_1mm(v: &Volume<f32>) -> Result<Volume<f32>, String> {
    let (data, dims, affine) = (&v.data, &v.dims, &v.affine);
    let factors: [f64; 3] = std::array::from_fn(|a| {
        (0..3)
            .fold(0.0, |s, r| s + affine[r][a] * affine[r][a])
            .sqrt()
    });
    if factors.iter().any(|&s| !(0.05..=20.0).contains(&s)) {
        return Err("Voxel spacing is outside the supported range (0.05–20 mm).".into());
    }
    // Match NumPy arange(start, stop, step), including its endpoint rounding.
    let starts = factors.map(|f| -(f - 1.0) / (2.0 * f));
    let steps = factors.map(|f| 1.0 / f);
    let shape: [usize; 3] = std::array::from_fn(|a| {
        (((starts[a] + steps[a] * (dims[a] as f64 * factors[a]).ceil()) - starts[a]) / steps[a])
            .ceil() as usize
    });
    if product(&shape) > 48 * 1024 * 1024 {
        return Err(
            "The 1 mm output is too large. Crop excess background before synthesis.".into(),
        );
    }
    let filtered = gaussian(
        data,
        dims,
        factors.map(|f| if f > 1.0 { 0.0 } else { 0.25 / f }),
    );
    let coords: Vec<Vec<f64>> = (0..3)
        .map(|a| {
            (0..shape[a])
                .map(|i| {
                    (starts[a] + i as f64 * ((starts[a] + steps[a]) - starts[a]))
                        .max(0.0)
                        .min((dims[a] - 1) as f64)
                })
                .collect()
        })
        .collect();
    let mut out = vec![0f32; product(&shape)];
    for z in 0..shape[2] {
        let fz = coords[2][z];
        let z0 = fz.floor() as usize;
        let z1 = (z0 + 1).min(dims[2] - 1);
        let tz = fz - z0 as f64;
        for y in 0..shape[1] {
            let fy = coords[1][y];
            let y0 = fy.floor() as usize;
            let y1 = (y0 + 1).min(dims[1] - 1);
            let ty = fy - y0 as f64;
            let b00 = dims[0] * (y0 + dims[1] * z0);
            let b10 = dims[0] * (y1 + dims[1] * z0);
            let b01 = dims[0] * (y0 + dims[1] * z1);
            let b11 = dims[0] * (y1 + dims[1] * z1);
            for x in 0..shape[0] {
                let fx = coords[0][x];
                let x0 = fx.floor() as usize;
                let x1 = (x0 + 1).min(dims[0] - 1);
                let tx = fx - x0 as f64;
                let lerp =
                    |b: usize| filtered[b + x0] as f64 * (1.0 - tx) + filtered[b + x1] as f64 * tx;
                let (v00, v10, v01, v11) = (lerp(b00), lerp(b10), lerp(b01), lerp(b11));
                out[index(x, y, z, &shape)] = ((v00 * (1.0 - ty) + v10 * ty) * (1.0 - tz)
                    + (v01 * (1.0 - ty) + v11 * ty) * tz)
                    as f32;
            }
        }
    }
    let mut aff = *affine;
    for r in 0..3 {
        for a in 0..3 {
            aff[r][a] /= factors[a];
        }
        aff[r][3] -= (0..3).fold(0.0, |v, a| v + aff[r][a] * 0.5 * (factors[a] - 1.0));
    }
    Ok(Volume {
        data: out,
        dims: shape,
        affine: aff,
    })
}

pub struct Prepared {
    pub input: Vec<f32>,
    pub padded: [usize; 3],
    pub aligned: [usize; 3],
    pub offsets: [usize; 3],
    pub axes: [usize; 3],
    pub flips: [bool; 3],
    pub dims: [usize; 3],
    pub affine: Affine,
}

pub fn prepare(v: &Volume<f32>, ct: bool) -> Result<Prepared, String> {
    let r = resample_1mm(v)?;
    let axes = ras_axes(&r.affine)?;
    let flips: [bool; 3] = std::array::from_fn(|i| r.affine[i][axes[i]] < 0.0);
    let aligned = axes.map(|a| r.dims[a]);
    let padded = aligned.map(|s| s.div_ceil(32) * 32);
    let offsets: [usize; 3] = std::array::from_fn(|a| (padded[a] - aligned[a]) / 2);
    let mut input = vec![0f32; product(&padded)];
    let has_padding = input.len() > r.data.len();
    let (mut min, mut max) = if has_padding {
        (0.0f64, 0.0f64)
    } else {
        (f64::INFINITY, f64::NEG_INFINITY)
    };
    for x in 0..aligned[0] {
        for y in 0..aligned[1] {
            for z in 0..aligned[2] {
                let mut src = [0usize; 3];
                for (a, p) in [x, y, z].into_iter().enumerate() {
                    src[axes[a]] = if flips[a] { aligned[a] - 1 - p } else { p };
                }
                let mut val = r.data[index(src[0], src[1], src[2], &r.dims)] as f64;
                if ct {
                    val = val.max(0.0).min(80.0);
                }
                input[((x + offsets[0]) * padded[1] + y + offsets[1]) * padded[2]
                    + z
                    + offsets[2]] = val as f32;
                min = min.min(val);
                max = max.max(val);
            }
        }
    }
    if max <= min {
        return Err("The input has no intensity variation after preprocessing.".into());
    }
    for v in input.iter_mut() {
        *v = ((*v as f64 - min) / (max - min)) as f32;
    }
    Ok(Prepared {
        input,
        padded,
        aligned,
        offsets,
        axes,
        flips,
        dims: r.dims,
        affine: r.affine,
    })
}

pub fn flip_input(input: &[f32], dims: &[usize; 3]) -> Vec<f32> {
    let slab = dims[1] * dims[2];
    let mut out = vec![0f32; input.len()];
    for x in 0..dims[0] {
        out[(dims[0] - 1 - x) * slab..(dims[0] - x) * slab]
            .copy_from_slice(&input[x * slab..(x + 1) * slab]);
    }
    out
}

pub fn finish(prediction: &[f32], p: &Prepared, sharpen: bool) -> Volume<u8> {
    let (d, pd, o) = (&p.aligned, &p.padded, &p.offsets);
    let mut aligned = vec![0f32; product(d)];
    for x in 0..d[0] {
        for y in 0..d[1] {
            for z in 0..d[2] {
                aligned[index(x, y, z, d)] =
                    prediction[((x + o[0]) * pd[1] + y + o[1]) * pd[2] + z + o[2]];
            }
        }
    }
    if sharpen {
        let smooth = gaussian(&aligned, d, [1.5; 3]);
        for (a, s) in aligned.iter_mut().zip(&smooth) {
            *a = (2.0 * *a as f64 - *s as f64) as f32;
        }
    }
    let mut data = vec![0u8; product(&p.dims)];
    for x in 0..d[0] {
        for y in 0..d[1] {
            for z in 0..d[2] {
                let mut dst = [0usize; 3];
                for (a, v) in [x, y, z].into_iter().enumerate() {
                    dst[p.axes[a]] = if p.flips[a] { d[a] - 1 - v } else { v };
                }
                // numpy uint8 conversion truncates, it does not round.
                data[index(dst[0], dst[1], dst[2], &p.dims)] =
                    (2.0 * aligned[index(x, y, z, d)] as f64)
                        .max(0.0)
                        .min(255.0) as u8;
            }
        }
    }
    Volume {
        data,
        dims: p.dims,
        affine: p.affine,
    }
}
